### ISSUE TITLE
rust-analyzer shows false positive E0308 type mismatch errors with PyO3 macros

### ISSUE THREAD

**Issue by sir-stevie** [rust-lang/rust-analyzer#20068]
## Description

rust-analyzer shows false positive E0308 type mismatch errors when using PyO3's `#[pyclass]` macro in combination with third-party procedural macros (specifically `pyo3_stub_gen_derive::gen_stub_pyclass`). The errors appear as:

```error
expected &'static [fn() -> TypeInfo], found &[fn() -> TypeInfo; 0]
expected &'static [MemberInfo], found &[MemberInfo; 0]
expected &'static [PyType_Slot], found &[PyType_Slot; 0]
expected &'static [MaybeRuntimePyMethodDef], found &[MaybeRuntimePyMethodDef; 0]
```

Despite these rust-analyzer errors, the code compiles and runs perfectly.

## Steps to Reproduce

1. Create a PyO3 project with the following dependencies:

```toml
[dependencies]
pyo3 = { version = "*" }
pyo3-polars = { version = "0.22.0", features = ["derive", "dtype-full", "lazy"] }
pyo3-stub-gen = { version = "0.10.0", default-features = false }
```

2. Create a struct with both PyO3 and stub generation macros:

```rust
use pyo3::prelude::*;
use pyo3_stub_gen::derive::{gen_stub_pyclass, gen_stub_pymethods};

#[gen_stub_pyclass]
#[pyclass]
#[derive(Clone)]
pub struct YourStruct;

#[gen_stub_pymethods]
#[pymethods]
impl YourStruct {
    #[staticmethod]
    fn example_method() -> PyResult<f64> {
        Ok(42.0)
    }
}
```

3. Open the project in VS Code with rust-analyzer extension
4. Observe the E0308 errors in rust-analyzer, while `cargo check` passes without issues

## Expected Behavior

rust-analyzer should not show type mismatch errors for properly written PyO3 code that compiles successfully.

## Actual Behavior

rust-analyzer shows multiple E0308 type mismatch errors related to array-to-slice coercion in macro-generated code.

## Environment

- **Rust version**: 1.89.0 (nightly)
- **rust-analyzer version**: v0.3.2509
- **Operating System**: macOS
- **Python version**: 3.13
- **Cargo.toml setup**:

```toml
  pyo3 = { version = "*" }
  pyo3-polars = { version = "0.22.0", features = ["derive", "dtype-full", "lazy"] }
  pyo3-stub-gen = { version = "0.10.0", default-features = false }
```

## Additional Context

- The issue appears to be related to how rust-analyzer handles procedural macro expansion when multiple proc macros are combined
- The code works perfectly at runtime and passes all `cargo` commands
- This affects developer experience as it creates noise in the IDE with false error indicators

## Workaround

Currently, disabling E0308 diagnostics in rust-analyzer settings as a workaround.

**Comment by Distortedlogic** [comment]
I believe there is a regression in one of the latest nightly versions .. downgrading to a few days ago resolves .. I seen this `'static` type error as well as a slice vs sized slice `[u8]` vs `[u8; 5]` in marco expansions with the nightly that has the regression

**Comment by ShoyuVanilla** [comment]
Strangely enough, I can't reproduce this. Could you please share your `Cargo.lock` on your reproduction steps for exact dependency versions as well, with exact rustc nightly version(with dates and hash) output that printed with `rustc --version`?

**Comment by sir-stevie** [comment]
> Strangely enough, I can't reproduce this. Could you please share your `Cargo.lock` on your reproduction steps for exact dependency versions as well, with exact rustc nightly version(with dates and hash) output that printed with `rustc --version`?

I shared everything in the description I gave about the setup I was using. I used @Distortedlogic 's trick to get around it, I'm not having the issue anymore. I'll see if a I ca reproduce it again and share the `Cargo.lock`

**Comment by Distortedlogic** [comment]
> Strangely enough, I can't reproduce this. Could you please share your `Cargo.lock` on your reproduction steps for exact dependency versions as well, with exact rustc nightly version(with dates and hash) output that printed with `rustc --version`?

did u use the nightly build of the day the issue was posted? like the june 22nd nightly 1.89? nightly is on 1.90 now.

https://github.com/rust-lang/rust/pull/142794
here is a rollup of 9 PRs which are all in the 1.89 nightly release change log and the timing matches

take this with a grain of salt as I am no rust pro... but I think the issue we saw was related to https://github.com/rust-lang/rust/issues/142532

The macro expansion where I hit this issue were using `unsafe` and pointer operations .. and the location of the pointer operation was where I was getting the type error of `&[u8]` vs `&[u; 5]`

https://github.com/rust-lang/rust/pull/142650 - this one seems related to rust-analyzer but over my head for any kinda guess on its impact

https://github.com/rust-lang/rust/pull/142713 - this one seems related to macros

**Comment by ensc** [comment]
I am seeing the E0308 when using `tracing` macros and/or in code like

```rust
pub trait Foo: Sync {}
struct Bar;
impl Foo for Bar {}

static BAR: Bar = Bar;
pub static BAR_PTR: &'static dyn Foo = &BAR;    // <<<< expected &'static (dyn Foo + 'static), found &Bar

fn foo() {
    // expected &'static (dyn Callsite + 'static), found &DefaultCallsite
    // expected &'static [&'static str], found &[&'static str; 1]
    tracing::error!("foo");
}
```

rust-analyzer 1.90.0-nightly (bdaba05 2025-06-27)

**UPDATE**: ok with rust-analyzer 1.90.0-nightly (71e4c00 2025-07-01)

**Comment by Veykril** [comment]
I believe this is due to the sized hierachy changes, latest rust-analyzer pre-release should be working again now

**Comment by qRoC** [comment]
More simple:

```rust
struct Test;

impl std::fmt::Debug for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // expected &(dyn Debug + 'static), found &&'static str
        f.debug_tuple("Test").field(&"hello").finish()
    }
}
```

rustc 1.90.0-nightly (bdaba05a9 2025-06-27) 
rust-analyzer 0.3.2509-standalone (0100bc7373 2025-06-22)

**Update**
Ok in: `rust-analyzer 0.4.2517-standalone (13ffda70eb 2025-06-28)`

**Comment by ShoyuVanilla** [comment]
> **Update** Ok in: `rust-analyzer 0.4.2517-standalone (13ffda70eb 2025-06-28)`

Good! Is there anyone getting this issue with the latest stable (or nightly) version of rust-analyzer?
If not, I think this is good to be closed and if so, I'll take a look into this.

**Comment by ensc** [comment]
Still with rust-analyzer 1.90.0-nightly (e43d139 2025-07-09)

```rust
use std::any::Any;
use std::sync::Arc;

pub trait Device: Any + Sync + Send {
}

pub struct Gpio;
impl Device for Gpio {}

pub fn ext_dev(dev: Arc<dyn Device>) -> Arc<Gpio> {
    Arc::downcast::<Gpio>(dev).unwrap()
    //                    ^^^
    // expected Arc<dyn Any + Send + Sync, Global>, found Arc<dyn Device, Global>
}
```

**Comment by ChayimFriedman2** [comment]
Works now.
